snapcraft/internal/meta.py

```diff
@@ -116,6 +119,38 @@ def setup_assets(self):
             file_utils.link_or_copy(
                 'gadget.yaml', os.path.join(self.meta_dir, 'gadget.yaml'))
 
+    def write_snap_directory(self):
+        # First migrate the snap directory. It will overwrite any conflicting
+        # files.
+        for root, directories, files in os.walk('snap'):
+            for directory in directories:
+                source = os.path.join(root, directory)
+                destination = os.path.join(self._snap_dir, source)
+                file_utils.create_similar_directory(source, destination)
+
+            for file_path in files:
+                source = os.path.join(root, file_path)
+                destination = os.path.join(self._snap_dir, source)
+                with contextlib.suppress(FileNotFoundError):
+                    os.remove(destination)
+                file_utils.link_or_copy(source, destination)
+
+        # Now write hook wrappers for any hooks included in that directory.
+        self._generate_hook_wrappers()
```

Comment:
Apps are related in that they're runnable things exposed by snapcraft, as are hooks. One is run by the user, one is run by snapd, but they're both exposed by snapcraft. What I mean to say is, if we don't expect developers to want to do such a thing in apps, why do we expect them to want to do such a thing in a hook?